File: lintcode/next-permutation.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/next-permutation/

class Solution:
    # @param num :  a list of integer
    # @return : a list of integer
    def nextPermutation(self, num):
        # write your code here
        if not num:
            return num

        for i in xrange(len(num)-2, -1, -1):
            if num[i+1] > num[i]:
                for j in xrange(len(num)-1, i-1, -1):
                    if num[j] > num[i]:
                        break
                num[i], num[j] = num[j], num[i]
                num[i+1:] = num[:i:-1]
                return num
        return num[::-1]

if __name__ == '__main__':
    s = Solution()
    print s.nextPermutation([1,3,2,3])
    print s.nextPermutation([1,6,5,4])
    print s.nextPermutation([4,3,2,1])

File: misc/google/next-permutation.cc
// Question: Next permutation.

void solve(vector<int> nums) {
    int len = nums.size();
    int i = len-2;
    while (i >= 0) {
        if (nums[i+1] > nums[i]) {
            int j = len-1;
            while (i < j) {
                if (nums[j] > nums[i]) {
                    break;
                }
                --j;
            }
            swap(nums[i], nums[j]);
            reverse(nums.begin() + i + 1, nums.end());
            return;
        }
        --i;
    }
    reverse(nums.begin(), nums.end());
}
